components/script/dom/offscreencanvas.rs

```diff
@@ -0,0 +1,178 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use dom::bindings::codegen::Bindings::OffscreenCanvasBinding::{OffscreenCanvasMethods, Wrap as OffscreenCanvasWrap, OffscreenRenderingContext};
+use dom::bindings::cell::DomRefCell;
+use dom::bindings::error::{Error, Fallible};
+use dom::bindings::reflector::reflect_dom_object;
+use dom::bindings::root::{DomRoot, Dom};
+use dom::bindings::str::DOMString;
+use dom::eventtarget::EventTarget;
+use dom::globalscope::GlobalScope;
+use dom::htmlcanvaselement::HTMLCanvasElement;
+use dom::offscreencanvasrenderingcontext2d::OffscreenCanvasRenderingContext2D;
+use dom_struct::dom_struct;
+use euclid::Size2D;
+use js::rust::HandleValue;
+use js::jsapi::JSContext;
+use ref_filter_map;
+use std::cell::Ref;
+use std::cell::Cell;
+
+#[must_root]
+#[derive(Clone, JSTraceable, MallocSizeOf)]
+pub enum OffscreenCanvasContext {
+    OffscreenContext2d(Dom<OffscreenCanvasRenderingContext2D>),
+    //WebGL(Dom<WebGLRenderingContext>),
+    //WebGL2(Dom<WebGL2RenderingContext>),
+}
+
+
+#[dom_struct]
+pub struct OffscreenCanvas{
+    eventtarget: EventTarget,
+    height: Cell<u64>,
+    width: Cell<u64>,
+    context: DomRefCell<Option<OffscreenCanvasContext>>,
+    placeholder: Option<Dom<HTMLCanvasElement>>,
+}
+
+impl OffscreenCanvas{
+    pub fn new_inherited(height: u64, width: u64, placeholder: Option<&HTMLCanvasElement>) -> OffscreenCanvas {
+        OffscreenCanvas {
+            eventtarget: EventTarget::new_inherited(),
+            height: Cell::new(height),
+            width: Cell::new(width),
+            context: DomRefCell::new(None),
+            placeholder: placeholder.map(Dom::from_ref),
+        }
+    }
+
+    pub fn new(global: &GlobalScope, height: u64, width: u64, placeholder: Option<&HTMLCanvasElement>) -> DomRoot<OffscreenCanvas> {
+        reflect_dom_object(Box::new(OffscreenCanvas::new_inherited(height,width,placeholder)), global, OffscreenCanvasWrap)
+    }
+
+    pub fn Constructor (global: &GlobalScope, height: u64, width: u64) -> Fallible<DomRoot<OffscreenCanvas>> {
+        //step 1
+        let offscreencanvas = OffscreenCanvas::new(global,height,width,None);
+        //step 2
+
+        if offscreencanvas.context.borrow().is_some() {
```

Comment:
There is no step 1 or step 2 in https://html.spec.whatwg.org/multipage/canvas.html#dom-offscreencanvas, and the spec does not include any check for the context. Please remove this and the commented code.